homeassistant/components/heos/__init__.py

```diff
@@ -219,6 +227,115 @@ def update_ids(self, mapped_ids: dict[int, int]):
+            member_entity_ids = [
+                player_id_to_entity_id_map[member.player_id]
+                for member in group.members
+                if member.player_id in player_id_to_entity_id_map
+            ]
+            # Make sure the group leader is always the first element
+            group_info = [leader_entity_id, *member_entity_ids]
+            group_info_by_entity_id[leader_entity_id] = group_info
+            for member_entity_id in member_entity_ids:
+                group_info_by_entity_id[member_entity_id] = group_info
+
+        return group_info_by_entity_id
+
+    async def async_join_players(
+        self, leader_entity_id: str, member_entity_ids: list[str]
+    ) -> None:
+        """Create a group with `leader_entity_id` as group leader and `member_entities` as member players."""
+        entity_id_to_player_id_map = self._get_entity_id_to_player_id_map()
+        leader_id = entity_id_to_player_id_map.get(leader_entity_id)
+        member_ids = [
+            entity_id_to_player_id_map.get(member) for member in member_entity_ids
+        ]
+
+        try:
+            await self.controller.create_group(leader_id, member_ids)
+        except HeosError as err:
+            _LOGGER.error(
+                "Failed to group %s with %s: %s",
+                leader_entity_id,
+                member_entity_ids,
+                err,
+            )
+
+    async def async_unjoin_player(self, player_entity_id: str):
+        """Remove `player_entity` from any group."""
+        player_id = self._get_entity_id_to_player_id_map().get(player_entity_id)
+
+        try:
+            await self.controller.create_group(player_id, [])
+        except HeosError as err:
+            _LOGGER.error(
+                "Failed to ungroup %s: %s",
+                player_entity_id,
+                err,
+            )
+
+    async def async_update_groups(self, event, data=None):
+        """Update the group membership from the controller."""
+        if event in (
+            heos_const.EVENT_GROUPS_CHANGED,
+            heos_const.EVENT_CONNECTED,
+        ):
+            groups = await self.async_get_group_membership()
+            if groups:
+                self._group_membership = groups
+                _LOGGER.debug("Groups updated due to change event")
+                # Let players know to update
+                self._hass.helpers.dispatcher.async_dispatcher_send(SIGNAL_HEOS_UPDATED)
+            else:
+                _LOGGER.debug("Groups empty")
+
+    def connect_update(self):
+        """Connect listener for when groups change and signal player update."""
+        self.controller.dispatcher.connect(
+            heos_const.SIGNAL_CONTROLLER_EVENT, self.async_update_groups
+        )
+        self.controller.dispatcher.connect(
+            heos_const.SIGNAL_HEOS_EVENT, self.async_update_groups
+        )
+
+    def force_update_groups(self):
```

Comment:
Why do we need this method?

Reply:
The players need to call this when they have registered their `entity_id`. But since all other use-cases for calling this method have vanished now I guess it makes more sense to use a signal for this. So I have removed this method and the players are now sending `SIGNAL_HEOS_PLAYER_ADDED` which is handled by the group manager.